presto/dbapi.py

```diff
@@ -298,13 +299,27 @@ def _format_prepared_param(self, param):
         Formats parameters to be passed in an
         EXECUTE statement.
         """
+        if isinstance(param, bool):
+            return "true" if param else "false"
+        if isinstance(param, float):
+            return "DOUBLE '%s'" % param
         if isinstance(param, str):
             return ("'%s'" % param.replace("'", "''"))
-        else:
-            # TODO: param could have many other types which should
-            #       be covered including unicode, maps, lists, etc.
-            #       https://github.com/prestosql/presto-python-client/issues/45
-            return str(param)
+        if isinstance(param, bytes):
+            return "X'%s'" % param.hex()
+        if isinstance(param, datetime.datetime):
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z").rstrip(" ")
+            return "TIMESTAMP '%s'" % re.sub(r'\.([0-9]{3})([0-9]{3})', r'.\1', datetime_str)
+        if isinstance(param, list):
+            return "ARRAY[%s]" % ','.join(map(self._format_prepared_param, param))
+        if isinstance(param, dict):
+            keys = self._format_prepared_param(list(param.keys()))
+            values = self._format_prepared_param(list(param.values()))
```

Comment:
Can `param.keys()` and `param.values()` ordering go out of sync? can it depend on `dict` subclass?
if possible, this could be a safe guard:
```
entries = list(param. items())
keys = self._format_prepared_param([k for (k, v) in entries])
values = self._format_prepared_param([v for (k, v) in entries])
```

Reply:
Correct, they can be out of order, good catch.  I ended up using the keys to build the list of values.